src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -209,5 +222,73 @@ static void VerifyMemberOnlyApplyToTypesOrStrings (SymbolAnalysisContext context
 				context.ReportDiagnostic (Diagnostic.Create (DiagnosticDescriptors.GetDiagnosticDescriptor (DiagnosticId.DynamicallyAccessedMembersOnPropertyCanOnlyApplyToTypesOrStrings), member.Locations[0], member.GetDisplayName ()));
 			}
 		}
+
+		static void VerifyDamOnDerivedAndBaseMethodsMatch (SymbolAnalysisContext context)
+		{
+			if (context.Symbol.TryGetOverriddenMember (out var overriddenSymbol) && overriddenSymbol is IMethodSymbol overriddenMethod
+				&& context.Symbol is IMethodSymbol method) {
+				VerifyDamOnMethodsMatch (method, overriddenMethod, context);
+			}
+		}
+
+		static void VerifyDamOnMethodsMatch (IMethodSymbol method, IMethodSymbol overriddenMethod, SymbolAnalysisContext context)
+		{
+			if (FlowAnnotations.GetMethodReturnValueAnnotation (method)  != FlowAnnotations.GetMethodReturnValueAnnotation (overriddenMethod))
+				context.ReportDiagnostic (Diagnostic.Create (
+					DiagnosticDescriptors.GetDiagnosticDescriptor (DiagnosticId.DynamicallyAccessedMembersMismatchOnMethodReturnValueBetweenOverrides),
+					method.Locations[0], method.GetDisplayName (), overriddenMethod.GetDisplayName ()));
+
+			for (int i = 0; i < method.Parameters.Length; i++) 
+				if (FlowAnnotations.GetMethodParameterAnnotation(method.Parameters[i]) != FlowAnnotations.GetMethodParameterAnnotation(overriddenMethod.Parameters[i]))
+					context.ReportDiagnostic (Diagnostic.Create (
+						DiagnosticDescriptors.GetDiagnosticDescriptor (DiagnosticId.DynamicallyAccessedMembersMismatchOnMethodParameterBetweenOverrides),
+						method.Parameters[i].Locations[0],
+						method.Parameters[i].GetDisplayName (), method.GetDisplayName (), overriddenMethod.Parameters[i].GetDisplayName (), overriddenMethod.GetDisplayName ()));
+
+			for (int i = 0; i < method.TypeParameters.Length; i++)
+				if (method.TypeParameters[i].GetDynamicallyAccessedMemberTypes () != overriddenMethod.TypeParameters[i].GetDynamicallyAccessedMemberTypes ())
+					context.ReportDiagnostic (Diagnostic.Create (
+						DiagnosticDescriptors.GetDiagnosticDescriptor (DiagnosticId.DynamicallyAccessedMembersMismatchOnGenericParameterBetweenOverrides),
+						method.TypeParameters[i].Locations[0],
+						method.TypeParameters[i].GetDisplayName (), method.GetDisplayName (),
+						overriddenMethod.TypeParameters[i].GetDisplayName (), overriddenMethod.GetDisplayName ()));
+
+			if (!method.IsStatic && method.GetDynamicallyAccessedMemberTypes () != overriddenMethod.GetDynamicallyAccessedMemberTypes ())
+				context.ReportDiagnostic (Diagnostic.Create (
+					DiagnosticDescriptors.GetDiagnosticDescriptor (DiagnosticId.DynamicallyAccessedMembersMismatchOnImplicitThisBetweenOverrides),
+					method.Locations[0],
+					method.GetDisplayName (), overriddenMethod.GetDisplayName ()));
```

Comment:
Could you please file an issue about the fact that linker doesn't validate that method overrides have the same annotation on the method itself? And also add tests for this case - it's not really covered anywhere as far as I can tell.

Reply:
I might be misunderstanding what you mean but it looks like we do have a test for this [here](https://github.com/dotnet/linker/blob/26e0c5cdc29a5a4f541100148b5cb7a814773110/test/Mono.Linker.Tests.Cases/DataFlow/VirtualMethodHierarchyDataflowAnnotationValidation.cs#L606-L633), and the linker reports this warning [here](https://github.com/dotnet/linker/blob/26e0c5cdc29a5a4f541100148b5cb7a814773110/src/linker/Linker.Dataflow/FlowAnnotations.cs#L540-L542).

Reply:
Sorry - you're right. I looked through the callers of `LogValidationWarning` and could not find the place where we would check on the method itself. But it's "hidden" in the `ValidateMethodParametersHaveNoAnnotations` - this method considers `this` as a parameter and that will validate the method-level annotations.

Please ignore my comment.